src/utils/reviewAlert.ts

```diff
@@ -1,50 +1,95 @@
+import { IS_IOS } from '@/env';
+import { logger } from '@/logger';
+
+const { RainbowRequestReview, RNReview } = NativeModules;
 
 export const AppleReviewAddress =
   'itms-apps://itunes.apple.com/us/app/appName/id1457119021?mt=8&action=write-review';
+export const PlayStoreAddress =
+  'market://details?id=me.rainbow&showAllReviews=true';
+
+const TWO_MONTHS = 1000 * 60 * 60 * 24 * 60;
+
+export const numberOfTimesBeforePrompt: {
+  [key in ReviewPromptAction]: number;
+} = {
+  UserPrompt: 0, // this should never increment
+  TimesLaunchedSinceInstall: 5,
+  SuccessfulFiatToCryptoPurchase: 3,
+  DappConnections: 4,
+  Swap: 3,
+  BridgeToL2: 1,
+  AddingContact: 1,
+  EnsNameSearch: 1,
+  EnsNameRegistration: 1,
+  WatchWallet: 1,
+  NftFloorPriceVisit: 1,
+};
 
-export const REVIEW_DONE_KEY = 'AppStoreReviewDone';
-export const REVIEW_ASKED_KEY = 'AppStoreReviewAsked';
-let reviewDisplayedInTheSession = false;
-const TWO_MONTHS = 2 * 30 * 24 * 60 * 60 * 1000;
+export const handleReviewPromptAction = async (action: ReviewPromptAction) => {
+  logger.info(`handleReviewPromptAction: ${action}`);
 
-export default async function maybeReviewAlert() {
-  const reviewAsked = await AsyncStorage.getItem(REVIEW_ASKED_KEY);
-  if (Number(reviewAsked) > Date.now() - TWO_MONTHS) {
+  if (action === ReviewPromptAction.UserPrompt) {
+    promptForReview();
     return;
   }
 
-  const reviewDone = await AsyncStorage.getItem(REVIEW_DONE_KEY);
-  if (reviewDone) {
+  const hasReviewed = ls.review.get(['hasReviewed']);
+  if (hasReviewed) {
     return;
   }
-  // update to prevent double showing alert in one session
-  if (reviewDisplayedInTheSession) {
+
+  const actions = ls.review.get(['actions']) || [];
+  const actionToDispatch = actions.find(a => a.id === action);
+  if (!actionToDispatch) {
     return;
   }
-  reviewDisplayedInTheSession = true;
 
-  AsyncStorage.setItem(REVIEW_ASKED_KEY, Date.now().toString());
+  const timeOfLastPrompt = ls.review.get(['timeOfLastPrompt']) || 0;
+  logger.info(`timeOfLastPrompt: ${timeOfLastPrompt}`);
+
+  actionToDispatch.numOfTimesDispatched += 1;
+  logger.info(`numOfTimesDispatched: ${actionToDispatch.numOfTimesDispatched}`);
 
+  ls.review.set(['actions'], actions);
+
+  if (
+    actionToDispatch.numOfTimesDispatched >=
+      numberOfTimesBeforePrompt[action] &&
+    timeOfLastPrompt + TWO_MONTHS <= Date.now()
+  ) {
+    logger.info(`Prompting for review`);
+
+    actionToDispatch.numOfTimesDispatched = 0;
+    ls.review.set(['actions'], actions);
+    ls.review.set(['timeOfLastPrompt'], Date.now());
+    promptForReview();
+  }
+};
+
+export const promptForReview = async () => {
   Alert.alert(
     lang.t('review.alert.are_you_enjoying_rainbow'),
     lang.t('review.alert.leave_a_review'),
     [
       {
         onPress: () => {
-          AsyncStorage.setItem(REVIEW_DONE_KEY, 'true');
-          if (reviewDisplayedInTheSession) {
-            return;
+          ls.review.set(['hasReviewed'], true);
```

Comment:
To my knowledge there isn't a way to get whether or not a user *actually* reviews, so this seems like the best way to mark this.